feat: add support for dynamic row height

jamesonhill opened the pull request:
This adds a new prop `estimatedRowHeight`, which updates the Grid component to use `VariableSizeGrid` from `react-window` instead of `FixedSizeGrid`. When `estimatedRowHeight` is provided, `rowHeight` is ignored and content is dynamically sized. `estimatedRowHeight` is used as the default height while measuring the content.

nihgwu commented:
Thanks for your contribution, that's a great feature, I have some questions before I take some time for a detailed review
* this package is supposed to support React@^16.0.0 while you are using some new api only available in 16.3.0, that would be a breaking change(although I don't think that would be a big problem)
* Have you tested with frozen columns, in that mode the there would be 2 or 3 tables actually to achieve the frozen feature, so I think you need to check if the current height is the largest when setting the row height

jamesonhill commented:
@nihgwu Hi, yes it does currently work w/ frozen columns as well as frozen rows. LeftGrid, RightGrid, and MainGrid each call `setRowHeight`, https://github.com/Autodesk/react-base-table/pull/170/files#diff-1ce5af19cf8e0ddaa9ce889d8ac30d42R455, and if the key already exists and new size is bigger it updates the size. I am going to push one small update now so that it's not directly mutating the state object though.

nihgwu commented on src/BaseTable.js:
```diff
@@ -315,7 +316,10 @@ class BaseTable extends React.PureComponent {
     const cellProps = { isScrolling, cellData, columns, column, columnIndex, rowData, rowIndex, container: this };
-    const cell = renderElement(cellRenderer || <TableCell className={this._prefixClass('row-cell-text')} />, cellProps);
+    const cellClasses = cn(this._prefixClass('row-cell-text'), {
+      [this._prefixClass('row-cell-text-dynamic')]: this.props.useDynamicRowHeight,
```
I think the dynamic should be promoted to the row(`row--dynamic`) and use `:not(.row--dynamic) { .cell-text{} }`

nihgwu replied:
perhaps we should promote even further to the top level like `table--dynamic`, as all the rows would have the same `row--dynamic` className

jamesonhill replied:
I was thinking maybe only applying `row-cell-text` when not in dynamic mode.  None of those properties have any affect when content is dynamically sized, so always adding them and overriding is un-necessary work.

nihgwu commented on src/BaseTable.js:
```diff
@@ -445,8 +449,22 @@ class BaseTable extends React.PureComponent {
     );
   }
 
+  setRowHeight = (index, size) => {
```
we don't use arrow functions in a class component

nihgwu replied:
Have you considered about expansion? The indices would change on expand/collapse

jamesonhill replied:
Expansion seems to have fine performance for me locally. I can update this to not use arrow functions.

jamesonhill replied:
I think i see what you were alluding to w/ expansion. It seems its not entirely correct. I'll dig into fixing it.

nihgwu replied:
I think you can use `rowKey` as the key instead of `index`, as you can have totally different data with the same index

nihgwu commented on src/BaseTable.js:
```diff
@@ -445,8 +449,22 @@ class BaseTable extends React.PureComponent {
+
+    if (rowHeightMap.hasOwnProperty(index)) {
+      if (size > rowHeightMap[index]) {
+        rowHeightMap[index] = size;
```
If you are using `index` as the key, why not use an array directly?
also you should use `{...map, [index]: size}` to mutate the ref
and if there is no need to re-render, you don't need to always setState at the bottom

jamesonhill replied:
I didn't use an array bc of how `rowIndex` is negative for frozen rows. It seemed simpler to pass `props.rowIndex` directly through instead of adding logic to figure out original index.

nihgwu replied:
yes, I totally forget that 👍

nihgwu commented on src/TableRow.js:
```diff
@@ -11,6 +13,17 @@ class TableRow extends React.PureComponent {
+    this.mounted = false;
+  }
+
+  componentDidMount() {
```
wha't if the rowData changed

jamesonhill replied:
ah, good point. I will add another check for when rowData change to re-measure the row.

jamesonhill commented:
@nihgwu As far as using `createContext` from React @16.3, I considered including that bump here but wasn't sure if it was appropriate or not. I can add it if you think so. It would be really fantastic if it could eventually get bumped to React@16.8 so that we can use hooks for new features.

nihgwu commented:
Yes in the next major version I will use hooks, but for this version we should try to avoid breaking changes

nihgwu commented:
For the context, I remember there is a compatible package there?

jamesonhill commented:
Perhaps. I haven't used any but I can check to see if one exists.

nihgwu commented:
I haven't check if we do need the new context, but here is the polyfill for it https://github.com/jamiebuilds/create-react-context

nihgwu commented on src/BaseTable.js:
```diff
@@ -1168,6 +1204,7 @@ BaseTable.propTypes = {
     ExpandIcon: PropTypes.func,
     SortIndicator: PropTypes.func,
   }),
+  useDynamicRowHeight: PropTypes.bool,
```
maybe we could use `estimatedRowHeight` for initial height and scrollToRow

nihgwu replied:
if this value is presented, we use `VariableSizeGrid` and ignore `rowHeight`

nihgwu commented on src/GridTable.js:
```diff
@@ -43,6 +45,37 @@ class GridTable extends React.PureComponent {
+    }
+
+    if (this.bodyRef) {
+      this.bodyRef.resetAfterRowIndex(0);
```
`resetAfterRowIndex(index)`?

jamesonhill replied:
`react-window` maintains an internal cache of row heights. Since we first use default row height prior to measuring, then measure content height, we have to call `resetRowAfterIndex` to clear that internal cache and use the updated heights passed to `Grid.renderRow`.

nihgwu replied:
I understand that, I'm just asking if we should only reset from the index instead of 0

jamesonhill replied:
oh i see. Yes let me try that.

jamesonhill replied:
Ok, so I remember now why I used 0 here instead of `rowIndex`. Since `Grid` is rendered before `Header`, without resetting the cache the increase in `Header` total height will be offset between `Header` and `Grid`.  I think here it does make sense to compute the original row index so as to avoid resetting more heights than necessary, so i'll do that.

nihgwu commented on src/GridTable.js:
```diff
@@ -1,16 +1,18 @@
  */
-class GridTable extends React.PureComponent {
+export const RowHeightContext = React.createContext({});
+class GridTable extends React.Component {
```
why not use pure component here?

jamesonhill replied:
I actually saw better rendering performance using `Component` over `PureComponent`.  With `PureComponent` there was small lag is resizing row height, which I didn't observe using `Component`.

nihgwu replied:
That's interesting, I'll have a look later

jamesonhill replied:
No need, i'll revert it back to PureComponent. After some of the recent changes I made i'm not seeing the lag anymore.

nihgwu commented on src/GridTable.js:
```diff
@@ -161,7 +209,8 @@ GridTable.propTypes = {
   bodyWidth: PropTypes.number.isRequired,
-  rowHeight: PropTypes.number.isRequired,
+  rowHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.func]).isRequired,
+  columnWidth: PropTypes.number,
```
I didn't see `columnWidth` is used anywhere

jamesonhill replied:
oops. Forgot to clean it up. I will remove.

nihgwu commented on src/TableRow.js:
```diff
@@ -11,6 +13,17 @@ class TableRow extends React.PureComponent {
+      const height = this.ref.current.getBoundingClientRect().height;
+      this.context.setRowHeightMap(rowIndex, height);
+    }
+    this.mounted = true;
```
I think we don't need context here, we can expose a `onRowHeightChange(index)` and use it in `BaseTable.renderRow`, in that way we don't need pass `setRowHeight` to the `TableRow` as well

jamesonhill replied:
removed context

nihgwu commented on src/TableRow.js:
```diff
@@ -55,11 +69,33 @@ class TableRow extends React.PureComponent {
+          this.context = context;
+          return (
+            <Tag style={style}>
+              <div
```
is there a way to eliminate an extra div node here, I understand it's necessary for size measurement, but I think we can wrap a div only if we are measuring? I thought about this problem before, another option would be measuring each cell, but that would be too expensive

jamesonhill replied:
hmm. Maybe? Originally I ran into issue measuring because if you set a height in css, then measuring will always return the initial height. I can keep messing around to see if there's another option though.

jamesonhill replied:
I was able to remove the extra node.

nihgwu commented on src/BaseTable.js:
```diff
@@ -473,14 +499,18 @@ class BaseTable extends React.PureComponent {
         rowRenderer={this.renderRow}
         onScroll={this._handleScroll}
         onRowsRendered={this._handleRowsRendered}
+        useDynamicRowHeight={useDynamicRowHeight}
```
I think we can use `rowHeight: func|number` here, then we don't need to pass so many extra props

jamesonhill replied:
Do you think `rowHeight` should become optional then? If user provides number or func, we use that otherwise it's internally calculated using the `TableRow.getBoundingClientRect().height` that I added?

nihgwu replied:
I think that's not the same case, for the `BaseTable`, the `rowHeight` could be optional as there is a default value, so actually we can remove the `isRequired` in the propTypes definition without breaking anything, and if we provide `estimatedRowHeight` we will ignore `rowHeight` and use `estimatedRowHeight` instead

for the `GridTable`, we use `rowHeight: number` as before, but `rowHeight: func` in dynamic mode, like `rowHeight = index => rowHeightMap[index]`, in this way we don't need to change the props interface at all

jamesonhill replied:
Ok, I think i'm following. I'll push an update and see if we're on same page.

nihgwu commented:
For get to mention that if you have tested with the infinite loading feature, I think we can use the innerRef to get the totalRowHeight instead of calculate from the count and rowHeight https://github.com/Autodesk/react-base-table/blob/master/src/BaseTable.js#L150, in that way I think we can make that feature works

jamesonhill commented:
@nihgwu I just pushed changes addressing your feedback. Freeze rows/columns, infinite loading, etc. all work & are backwards compatible. The one thing that doesn't work 100% perfect only when in dynamic mode is expansion. Since it currently passes `rowIndex` as the key, the expanded columns size to the parent row. I thought about setting `rowIndex-depth` as the key, however `Grid` doesn't know about `depth`. Do you have any ideas on how to make expansion correct when in dynamic?

nihgwu commented:
@jamesonhill thanks so much for your rapid response, I'll take a look at the the expansion later, and do you mind if I commit directly to your branch to make some minor changes if possible

jamesonhill commented:
@nihgwu yes, if you can make improvements you are welcome to commit directly on this branch.

nihgwu commented:
just realize that it doesn't work with resizable columns, I can help with this scenario, I'd like to change the resize behavior to only resize the cells on release instead of during that

nihgwu commented:
I just want to thank you again for your contribution, it's not an easy task when you want to make it a feature as we have to consider other scenarios as well, but we are almost there, thank you so much

nihgwu commented on src/BaseTable.js:
```diff
@@ -64,13 +64,17 @@ class BaseTable extends React.PureComponent {
 
     this._setContainerRef = this._setContainerRef.bind(this);
     this._setMainTableRef = this._setMainTableRef.bind(this);
+    this._setMainTableInnerRef = this._setMainTableInnerRef.bind(this);
```
I think we don't need those refs actually, we can move `getTotalRowsHeight` to the `GridTable` and use `this.table.getTotalRowsHeight` directly as all the three tables should have the same height

nihgwu commented on src/BaseTable.js:
```diff
@@ -455,11 +493,14 @@ class BaseTable extends React.PureComponent {
       // make sure `scrollLeft` is always integer to fix a sync bug when scrolling to end horizontally
       tableWidth = Math.max(Math.round(columnsWidth), tableWidth);
     }
+    const cls = cn(this._prefixClass('table-main'), {
```
I'm sorry for the misleading, I meant to add it to https://github.com/Autodesk/react-base-table/blob/master/src/BaseTable.js#L631

jamesonhill replied:
ah, i see. Yes I can add it there.

nihgwu commented on src/GridTable.js:
```diff
@@ -68,28 +84,40 @@ class GridTable extends React.PureComponent {
-          height={Math.max(height - headerHeight - frozenRowsHeight, 0)}
-          rowHeight={rowHeight}
+          height={Math.max(height - headerHeight - frozenRowsHeight(), 0)}
+          rowHeight={useDynamicRowHeight ? this.getRowHeight : rowHeight}
```
we need to pass `estimatedRowHeight` to the grid as well to make sure the `VariableSizeGrid` initialize the correct inner height

jamesonhill replied:
yes this is true. Good catch.

nihgwu commented on src/GridTable.js:
```diff
@@ -176,6 +204,9 @@ GridTable.propTypes = {
   rowRenderer: PropTypes.func.isRequired,
+  rowHeightMap: PropTypes.object,
+  estimatedRowHeight: PropTypes.number,
+  innerRef: PropTypes.func,
```
if you move `getTotalRowsHeight` here you don't need this prop anymore

jamesonhill commented:
No problem. I'm happy to help. This library is going to be very useful for me. It's the least I can do.

nihgwu commented on src/BaseTable.js:
```diff
@@ -445,8 +463,30 @@ class BaseTable extends React.PureComponent {
   }
 
+  setRowHeight(rowKey, size) {
+    this.setState(prev => {
```
I believe we could have other options, for example we could distinguish the calls in the same batch/rendering, I'll think about it

jamesonhill replied:
I think I got a little performance boost w/ your suggestion of calling setState less often. Previously scrolling to row 100 would result in 300 setState calls. Now only about 100.

jamesonhill commented:
Thanks for your help w/ expansion.

nihgwu commented:
![resize](https://user-images.githubusercontent.com/2595058/84519129-393d1400-ad04-11ea-937a-509d3f8f6f8f.gif)
it works with resizing like a charm 🎉

jamesonhill commented:
@nihgwu Looks awesome! Nice work.

nihgwu commented:
it's kinda hacky to make it work with frozen data, I'll do more test before land this feature

jamesonhill commented:
@nihgwu do u have any un-pushed config changes? Since yesterday the `website` project won't build for me. The error `Can't import the named export 'array' from non EcmaScript module (only default export is available` seems too indicate some kinda webpack issue, but from what I understand Gatsby manages webpack internally (though i'm not sure familiar w/ gatsby other than basics).

nihgwu commented:
no, everything is pushed to upstream

nihgwu commented:
@jamesonhill Have you resolved the problem? Have you tried to run `yarn` in the `website` folder as I added a new package `faker`

jamesonhill commented:
@nihgwu yea, i ended up wiping out my node_modules & re-installing all of the deps, which resolved it. I should have some time tomorrow to do some more testing to make sure all features are working. Seems like we're pretty close.

nihgwu commented:
@jamesonhill Cool, it's a long-waited and most powerful feature, I never thought it would be implemented in V1, thanks for the efforts on that

nihgwu commented:
Hi @jamesonhill I won't have enough time until the weekend, so if you can do some tests or further improvements that would be great, I'll try to get this PR merged in the weekend

jamesonhill commented:
@nihgwu sounds good to me. I'm  reviewing some of your recent changes and gonna see what else we can do.

jamesonhill commented:
There appears to be an issue w/ frozen columns. `onRowHover` is causing the height to shrink when it shouldn't. Still trying to track root issue down.

jamesonhill commented on src/TableRow.js:
```diff
@@ -10,9 +10,24 @@ class TableRow extends React.PureComponent {
+    this.props.estimatedRowHeight && this._measureHeight();
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    if (this.props.estimatedRowHeight && this.state.measured === prevState.measured) {
+      this.setState({ measured: false }, this._measureHeight);
+    }
+  }
```
This seems to be what is causing the frozenColumns issue. I think this may be creating an infinite loop of calling `measureHeight`

nihgwu replied:
Oh yes, I can repro the issue by set `fixed=true`, the latest commits were tested with `fixed=false` to work with column resizing

nihgwu replied:
I just pushed a new commit which fixe the loop, but there are still something to fix, but I don't have enough time right now
We have to re-measure the row on each render if we don't use `ResizeObserver`, but I add a new check to make sure the measure only happens if the re-render is not caused by `resetAfterIndex`

jamesonhill commented on src/BaseTable.js:
```diff
@@ -935,19 +937,20 @@ class BaseTable extends React.PureComponent {
     this.leftTable && this.leftTable.resetAfterRowIndex(rowIndex);
     this.rightTable && this.rightTable.resetAfterRowIndex(rowIndex);
-    this.props.frozenData && this.forceUpdate();
+    // this.props.frozenData && this.forceUpdate();
```
think we need to keep this

nihgwu commented:
Sure, but I’m thinking better better ways to handle that case, or do we really need dynamic frozen rows? Which makes to case too complicated

On Jun 17, 2020, at 02:28, Jameson Hill <notifications@github.com> wrote:

﻿

@jamesonhill comment on this pull request.

________________________________

In src/BaseTable.js<https://github.com/Autodesk/react-base-table/pull/170#discussion_r441058255>:

>    }

   _resetAfterRowIndex(rowIndex) {
     this.table && this.table.resetAfterRowIndex(rowIndex);
     this.leftTable && this.leftTable.resetAfterRowIndex(rowIndex);
     this.rightTable && this.rightTable.resetAfterRowIndex(rowIndex);
-    this.props.frozenData && this.forceUpdate();
+    // this.props.frozenData && this.forceUpdate();


think we need to keep this

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Autodesk/react-base-table/pull/170#pullrequestreview-431783630>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AATZR4RC5DJV3JCRQKVSVNTRW62UNANCNFSM4NYU5XLA>.

jamesonhill commented:
True, maybe frozen rows just keep a fixed height. I'll defer to your judgement on that. The only other issue i've seen since your recent changes are w/ resizing columns & dynamic row heights. The row container width isn't updating, so when you hover over a row theres an empty space unhighlighted. I'm working on figuring out how to fix that now.

jamesonhill commented:
If frozen rows have fixed height, maybe we should update their `.row-cell-text` to have `text-wrap: nowrap` like non-dynamic tables?

nihgwu commented on src/BaseTable.js:
```diff
@@ -918,6 +918,11 @@ class BaseTable extends React.PureComponent {
     if (!resizingKey || !resizingWidth) return;
 
+    if (this.props.estimatedRowHeight) {
+      this.leftTable && this.leftTable.resetAfterColumnIndex();
```
Cool, I'm planning to use `VariableSizeList` instead of `VariableSizeGrid`, I think that would fix this problem out of box, as we are actually use only one column, maybe in a following PR

nihgwu replied:
I don't do that before as there is scrollbar issue with the `List`, but I guess I can solve that by `innerElementType`

nihgwu commented:
@jamesonhill right, I just don't have enough time on that, tried to fix the problem with minimum efforts now, more work would be done in the weekends

jamesonhill commented:
@nihgwu sounds good to me. I'm not seeing any other issues though. I'll keep testing, but this seems to be in pretty good shape. Scroll performance is noticeably better. Props to u on that.

nihgwu commented:
@jamesonhill I just pushed another commit, now you can reset cache without losing performance on resizing, it looks much better now
P.S. it's me who added `shouldForceUpdate` to `react-window` 😁

jamesonhill commented:
I haven't been able to track down what's causing it yet, but i'm seeing a little bit of "weird-ness" the first time I hover over a row (using dynamic heights). It gets smaller while hovering, then goes back to the correct size `onMouseLeave` and it doesn't happen on re-hover.

nihgwu commented:
@jamesonhill I've noticed that problem too, will inspect it

nihgwu commented:
@jamesonhill I've tracked down the issue, but just don't know why ;(, when you hover on a row the first time, the `this.props.style === prevProps.style = false` for the mainTable but `true` for the frozen tables, then the row height in the main table won't be measured, but I just don't understand how it happens, after reset, the style should be from the cache in `react-window` and the compare check should be true

jamesonhill commented:
@nihgwu  yea, that is odd. I'm debugging it and it even looks like they are the same unless i'm missing something.

jamesonhill commented:
Oh, nevermind. It looks like first render is using the estimatedRowHeight values, which makes sense.

nihgwu commented:
@jamesonhill the object value is the same but the reference is different before/after reset cache

nihgwu commented:
@jamesonhill I've fixed the issue, and I think it's ready to be merged, please help to review the code I changed, thanks

jamesonhill commented:
@nihgwu it looks good to me. I've done some local testing and don't see the hover issue anymore. Scrolling w/ dynamic row heights is quite fast too! I think it's ready to merge.

nihgwu commented:
@jamesonhill Cool, let's land it, thank you so much 🎉

jamesonhill commented:
Awesome, thanks for your assistance & guidance. This should only require a minor version bump to the npm package?

nihgwu commented:
sure, I'm experimenting with List instead of Grid, if it's a easy switch I want to include it, or I'll release a new version soon

nihgwu commented:
v1.10.0 is released, I've make List work but it requires api change on `onScroll`, so that's do it in the next major version

nihgwu commented:
@jamesonhill I noticed an issue regarding the out sync of measurement in productionhttps://autodesk.github.io/react-base-table/examples/dynamic-row-heights, I guess it's caused by the fact that it's much faster in production that debounce of `_updateRowHeights` doesn't work well as expected, so it won't wait for the row in three tables be measured and updated in a single batch, if I enlarge the debounce delay, it works fine but there is noticeable lag, maybe I have use another mechanism instead of buffer

jamesonhill commented:
hmm, I'm actually seeing it locally but not in that production example.

nihgwu commented:
LOL, that's interesting, I'm trying to fix it now
